### ISSUE TITLE
question: Why do you release a patch version with breaking changes

### ISSUE THREAD

**Issue by kevincodecookies** [typestack/class-transformer#553]
### Description
patch release with a breaking change

**Minimal code-snippet showcasing the problem**
just use the package like described by https://docs.npmjs.com/cli/v6/using-npm/semver#caret-ranges-123-025-004

### Expected behavior
no breaking changes on patch lvl releases

### Actual behavior
the version 0.3.2 has breaking changes as it moves and refactors an exported type (ClassType to ClassConstructor) and changes the signature of @Transform decorator (@Transform(value => //do something) to @Transform(({value}) => //do something))

### Recommendation
only do breaking changes on major lvl or on minor lvl if we are still on 0.someNumber.someNumber so that people using "^" can use the package correctly. like described here : https://docs.npmjs.com/cli/v6/using-npm/semver#caret-ranges-123-025-004

**Comment by AckerApple** [comment]
It looks like changes are under way for this looking at the logs here. At this time npm still has breaking version

my recommendation to fix now to prevent further auto injecting of bad version is to:
Release a version 0.3.3 that reverts the breaking changes AND THEN release a new version called 0.4.0 with the new breaking changed

**Comment by takayukioda** [comment]
I'm up to the idea "not include breaking change in patch version" too.

@NoNameProvided I agree to your idea that "library user should check changelog before update" (mentioned in https://github.com/typestack/class-transformer/issues/547#issuecomment-760451441) but it seems that you are missing following scenario.

Let's say, user were using [socket-controllers](https://github.com/typestack/socket-controllers) in user service and [reflect-metadata](https://github.com/rbuckton/reflect-metadata) released breaking change in "0.1.14". Even if user lock `reflect-metadata`'s version to "0.1.13", breaking change still slip through because `socket-controllers` dependency is not been locked (`^0.1.13`).
That's why I believe it's better to not release any breaking changes in patch version. Even when major version is zero.

I hope this scenario is convincible enough to fix this breaking change.
I would love to hear your thoughts.

**Comment by AckerApple** [comment]
As a maintainer of 30 packages, I would:

- download v0.3.1 from npm
- Publish v0.3.1 as v0.3.3 to overwrite breaking changes
- Download v0.3.2 from npm
- Release v0.3.2 as v0.4.0

No code changes. Let’s all make a mental note of this course of action to assist any package that has released breaking changes (cause it’s never too late)

**Comment by WSanders31** [comment]
> As a maintainer of 30 packages, I would:
> 
> * download v0.3.1 from npm
> * Publish v0.3.1 as v0.3.3 to overwrite breaking changes
> * Download v0.3.2 from npm
> * Release v0.3.2 as v0.4.0
> 
> No code changes. Let’s all make a mental note of this course of action to assist any package that has released breaking changes as a PATCH version (cause it’s never too late)

As someone who has already updated our source to utilize the new changes from v0.3.2, I for one, am terrified.

**Comment by AckerApple** [comment]
@WSanders31 heads up of where we had issue AFTER SUCCESSFUL build:

Used to work:
```
@Transform(parseInt)
```

No build errors occurred. Went through unit tests and luckily 200 was expected but 400 thrown for something like “cannot parse object”

Had to change to:

```
@Transform((ob: TransformFnParams) => parseInt(ob.value))
```

**Comment by gerryburtenshaw** [comment]
Breaking changes should result in a Major Release version update (not minor, or, in this case, patch)
So, these changes should have been released as a v1.0.0

**Comment by AckerApple** [comment]
@gerryburtenshaw i saw you updated your comment. I didn’t really agree with it at first and now I disagree with it more

Throw out your rule book because a package with 572,670 weekly downloads shouldn’t lean on the excuse of being in version zero.

Second, a lot of coders do not agree with changing the MAJOR version number for every breaking change. It would have saved a lot of Angulars reputation if it was 2.11 instead of 11.0. In life I never take rules as being perfect

**Comment by NoNameProvided** [comment]
`0.4.0` has been released, `0.3.2` has been deprecated with a warning message: 
> This release contains a breaking change from 0.3.1 but otherwise functions properly. Please read the changelog on Github to learn more.

Closing this as solved.

**Comment by gerryburtenshaw** [comment]
@AckerApple - I totally agree with you - I too was expecting the v1+ behaviour and not 0.x.y environment - I was only parroting the standard (which was previously mentioned)

In terms of rules, if the tools are enforcing them, then you kinda have to work within those constraints. Either your dancing with the devil or doing to 2-step around him.
